### PR TITLE
Allow S3 uploads without a timeout for exports

### PR DESCRIPTION
**Pull request by noxer**
#### Summary
Under certain conditions (for example when creating an export) we want to be able to upload a file without a timeout. This PR adds a method to the S3 file backend but not the interface. Go allows us to "query" for the availability of the method by doing:

```go
type ContextWriter interface {
	WriteFileContext(context.Context, io.Reader, string) (int64, error)
}

if cw, ok := fileBackend.(ContextWriter); ok {
	cw.WriteFileContext(ctx, r, path)
} else {
	fileBackend.WriteFile(r, path)
}
```

That way not all backends need to implement the new method. The export job now tries to disable the timeout when creating the export archive.

```release-note
Exports to S3 no longer time out
```

### CODE REVIEW

**Review by isacikgoz — APPROVED** [review]
LGTM

File: shared/filestore/s3store.go [hunk 1]
```diff
@@ -377,22 +384,26 @@ func (b *S3FileBackend) WriteFile(fr io.Reader, path string) (int64, error) {
...
-	// used. Bifrost needs to run in HTTPS, which is not yet deployed.
-	if buf, ok := fr.(*bytes.Buffer); ok && !isCloud {
-		objSize = buf.Len()
+	} else {
+		// We pass an object size only in situations where bifrost is not
+		// used. Bifrost needs to run in HTTPS, which is not yet deployed.
+		switch t := fr.(type) {
+		case *bytes.Buffer:
+			objSize = int64(t.Len())
+		case *os.File:
```
Comment (agnivade) [inline shared/filestore/s3store.go:399]: good one

**Review by agnivade — COMMENTED** [review]
looks good! Possible to add a test?

File: shared/filestore/filesstore_test.go [hunk 2]
```diff
@@ -121,6 +124,91 @@ func (s *FileBackendTestSuite) TestReadWriteFile() {
...
+		var (
+			written int64
+			err     error
+		)
+
+		path := "tests/" + randomString()
+
+		r, w := io.Pipe()
+		go func() {
+			// close the writer after two seconds
```
Comment (agnivade) [inline shared/filestore/filesstore_test.go:195]: incorrect comment?

**Review by agnivade — APPROVED** [review]
Looks good. Just a small one on an incorrect comment.

File: shared/filestore/filesstore_test.go (hunk 2 above)
Comment (noxer) [reply shared/filestore/filesstore_test.go:195]: :facepalm: yeah, I had it as 2 second and then lowered the timeouts so the tests don't stall.

**Pull request merged**